### PR TITLE
Implement methods in journey class.

### PR DESCRIPTION
**Pull request by ak47na**
This PR implements the method in `Journey.java.` 
A future PR will add a servlet that allows couriers to view the journeys.

### CODE REVIEW

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 1]
```diff
@@ -181,7 +181,7 @@ public Object getProperty(String orderKeyStr, String propertyName) throws Entity
 
   /** 
    * Updates the status property of orders with the keyString from orderKeys list to status.
-   TODO
+   * TODO[ak47na]: use transactions for updating order state.
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Note: Just updating all orders in a single transaction is theoretically not enough, you'd want to update all orders + create the new journey all in a single transaction.

The code is not layed out in a way to make that easy (as these 2 things happen in different classes), though you might still be able to make it work.

File: src/main/java/com/google/sps/data/Journey.java [hunk 2]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
   /**
    * Returns true if the minimum time journey fits in the timeslot and false otherwise.
    */
-  public boolean findJourneyForTimeslot(DeliverySlot deliverySlot) throws ApiException, IOException, InterruptedException {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  public boolean findJourneyForTimeslot(DeliverySlot deliverySlot) throws ApiException, DataNotFoundException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = findOptimalTimePaths();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java:92]: Be careful int is 32 bits, so this will go kaboom if you ever try to have more waypoints (is that why your current limit is 25?).

In general in java you would just use BitSet (https://docs.oracle.com/javase/8/docs/api/java/util/BitSet.html), which has some safety nets in place so you don't shoot yourself in the foot and would make the code much clearer.

File: src/main/java/com/google/sps/data/Journey.java [hunk 3]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
...
 
   public void addOrder(String orderKey) {
     orderKeys.add(orderKey);
   }
 
+  /**
+   * Computes for each waypoint i, pointsAfter[i] = bitmask where bits set to 1 represent the
+   * indices of points that must be visited after point i
+   */
+  private ArrayList<Integer> getPointsAfterFromRestricitions() {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: nit: typo Restrictions*

File: src/main/java/com/google/sps/data/Journey.java [hunk 4]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
...
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
+   * is lastWaypoint; config is a bitmask where bits set to 1 represent currently visited waypoints
+   */
+  private ArrayList<ArrayList<Pair>> initializeBestTimeMatrix() throws ApiException, IOException, InterruptedException, DataNotFoundException {
+    ArrayList<ArrayList<Pair>> bestTime = new ArrayList<>();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    for (int config = 0; config <= (1 << numberOfWaypoints); ++config) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: I don't undestand what bestTime[1 << numberOfWaypoints] records here.

Per your comment: "all points with bits set to 1 in config" "config is a bitmask where bits set to 1 represent currently visited waypoints".

So for numberOfWaypoints = 2, 1 << numberOfWaypoints is 100.
1 - ??
0 - second waypoint doesn't need to be visited
0 - first waypoint doesn't need to be visited

So what does it mean to visit waypoint number (numberOfWaypoints + 1)?

**Review by anamariamilcu — APPROVED** [review]

File: src/main/java/com/google/sps/data/Journey.java [hunk 5]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
...
+      // add the config-th line to the matrix with numberOfWaypoints columns of Integer.MAX_VALUE
+      ArrayList<Pair> currentRow = new ArrayList<Pair>();
+
+      for (int waypointIndex = 0; waypointIndex <= numberOfWaypoints; ++ waypointIndex) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: You lost me again with the off by one error.
This iterates over numerOfWaypoints + 1 elements.

File: src/main/java/com/google/sps/data/Journey.java [hunk 6]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
...
+      currentState = prevState;
+    }
+    // add the start of the journey at the end of orderedWaypoints
+    orderedWaypoints.add(start);
+    return orderedWaypoints;
+  }
+
+  /**
+   * Creates and initializes bestTime matrix, a matrix where bestTime[config][lastWaypoint] = the
+   * minimum time to visit all points with bits set to 1 in config such that the last visited point
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java:152]: How are you representing Time as a pair?

By reading your implementation, I think you set the second element of the pair always to -1?

File: src/main/java/com/google/sps/data/Journey.java [hunk 7]
```diff
@@ -79,45 +87,181 @@ public int getNumberOfWaypoints() {
...
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
+    int minTime = bestTime.get((1 << numberOfWaypoints)).get(0).first;
+    for (int waypointIndex = 0; waypointIndex < numberOfWaypoints; ++waypointIndex) {
+      if (bestTime.get(allWaypointsConfig).get(waypointIndex).first < minTime) {
+        minTime = bestTime.get(allWaypointsConfig).get(waypointIndex).first;
+      }
+    }
+
+    long deliverySlotDurationInSeconds = Duration.between(deliverySlot.getStartTime().toInstant(),
+        deliverySlot.getEndTime().toInstant()).toMillis() / 1000;
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: You don't like getSeconds?
https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#getSeconds--

File: src/main/java/com/google/sps/data/Journey.java [hunk 8]
```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
...
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint] is the min
+   * time needed to visit all points with bits set to 1 in config such that the path ends in point
+   * lastWaypoint.
+   */
+  public ArrayList<ArrayList<Pair>> findOptimalTimePaths()  throws ApiException, DataNotFoundException, IOException, InterruptedException {
+    ArrayList<ArrayList<Pair>> bestTime = initializeBestTimeMatrix();
+    ArrayList<Integer> pointsAfter = getPointsAfterFromRestricitions();
+    int numberOfWaypoints = getNumberOfWaypoints();
+    int allWaypointsConfig = (1 << numberOfWaypoints) - 1;
+    // start with config = 1 so that at least 1 point is set to 1 in config
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: I don't understand this comment. What exactly is set to 1?

File: src/main/java/com/google/sps/data/Journey.java [hunk 9]
```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
...
+      waypointIndex += 1;
+    }
+
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint] is the min
+   * time needed to visit all points with bits set to 1 in config such that the path ends in point
+   * lastWaypoint.
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: Can you add some details here on how this computation is performed?

File: src/main/java/com/google/sps/data/Journey.java [hunk 10]
```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
...
+        // before the current point have already been visited.
+        if ((config & (1 << waypointIndex)) == 0 && ((pointsAfter.get(waypointIndex) & config) == 0)) {
+          // Set bit wayPointIndex to 1 in newConfig if waypointIndex is set to zero in config AND
+          // all restrictions for waypointIndex are satisfied
+          int newConfig = config | (1 << waypointIndex);
+          for (int lastWaypoint = 0; lastWaypoint < numberOfWaypoints; ++lastWaypoint) {
+            // the lastWaypoint visited in config must be set to 1
+            if ((config & (1 << lastWaypoint)) != 0) {
+              // add the time needed to get from lastWaypoint to waypointIndex to bestTime[config][lastWaypoint]
+              // TODO[ak47na]: handle case when
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: ?

File: src/main/java/com/google/sps/data/Journey.java [hunk 11]
```diff
@@ -79,45 +87,184 @@ public int getNumberOfWaypoints() {
...
+
+    long deliverySlotDurationInSeconds = Duration.between(deliverySlot.getStartTime().toInstant(),
+        deliverySlot.getEndTime().toInstant()).toMillis() / 1000;
+
+    return (minTime <= (int)deliverySlotDurationInSeconds);
   }
 
   /**
    * Computes and returns the optimal order in which points should be visited starting from the
    * last point to the fist one (the start Point).
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: I'm a bit confused why do you return from last to first.

Why not just return the right order (e.g. orderedWaypoints.reverse())?

File: src/main/java/com/google/sps/data/Journey.java (hunk 8 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: For config = 1, point with index 0 is set to 1. The comment should illustrate the fact that the line with config = 0 should not be updated.

File: src/main/java/com/google/sps/data/Journey.java (hunk 6 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java:152]: The second element of the pair is -1 in the initialization and then it is set to the waypoint that was visited previously.
E.g. bestTime[(1 << (waypoint1) + 1 << (waypoint2)][waypoint2] = Pair(T, waypoint1) means that the optimal time to visit waypoint1 and waypoint2 is T and waypoint1 was visited right before waypoint2.

File: src/main/java/com/google/sps/data/Journey.java (hunk 5 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: I changed it to <.

File: src/main/java/com/google/sps/data/Journey.java (hunk 4 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: I'm sorry that was a bug I missed: I changed it to bestTime[(1 << numberOfWaypoints) - 1] . Test passed because there was no test for which the first added point should be visited last. I will add one test for that.

For numberOfWaypoints = 2, allWaypointsConfig = 3 = 11(2) meaning that waypoints with indexes 0 and 1 must be visited. The indexes of waypoints are between 0 and numberOfWaypoint - 1.

File: src/main/java/com/google/sps/data/Journey.java (hunk 3 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: Fixed.

File: src/main/java/com/google/sps/data/Journey.java (hunk 11 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: I changed it to the right order.

File: src/main/java/com/google/sps/data/Journey.java (hunk 10 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: it was a TODO to handle the case when `getTimeInSecondsBetweenPoints` fails, but know it throws `ApiException`. I removed the comment.

File: src/main/java/com/google/sps/data/Journey.java (hunk 9 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: Done.

File: src/main/java/com/google/sps/data/Journey.java (hunk 7 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java]: I changed it to getSecond() thank you!

File: src/main/java/com/google/sps/data/Journey.java (hunk 2 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/Journey.java:92]: I'll have a look on that, thanks! 
The reason why the limit is 25 is because the time complexity for optimizing one journey is quite bad : O((2^(numberOfWaypoints)) * numberOfWaypoints)

**Review by ghost — APPROVED** [review]

File: src/main/java/com/google/sps/data/CourierStop.java [hunk 12]
```diff
@@ -33,6 +33,23 @@ public CourierStop(Point point) {
     orderKeys = new ArrayList<>();
   }
 
+  public Point getPoint() {
+    return point;
+  }
+
+  @Override
+  public boolean equals(Object object) {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/CourierStop.java:41]: Any reason we even support comparing to non-CourierStop objects?

That sounds like it might be a programming mistake, so failing at compile-time (code does not compile) would be preferable to silently continuing.

File: src/main/java/com/google/sps/data/Journey.java (hunk 6 above)
Comment (alexcepoi) [reply src/main/java/com/google/sps/data/Journey.java:152]: Thanks. Please work your explanation in the comment here (some of these comments are made not because I did not understand, but because I think this should be explained in the comment, rather than having the reader have to read the implementation to figure it out how this should work).

For future: In general, while sentinels (-1, MAX_VALUE) are very popular in programming contests, they are generally not very good in terms of safety checks/readability. null and java.util.Optional are more popular to represent 'not set' concepts (and force the caller to be explicit on how they handle these corner cases).

File: src/test/java/com/google/sps/data/JourneyHandlerTest.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * Tests that orders are correctly assigned to delivery slots.
+ */
+@RunWith(JUnit4.class)
+public class JourneyHandlerTest {
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+  private JourneyHandler journeyHandler;
+  private OrderHandler orderHandler;
+  private ManhattanDistancePathFinder pathFinder;
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/JourneyHandlerTest.java:53]: nit: this needn't be stored.

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 14]
```diff
@@ -152,16 +155,40 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
...
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity order = datastore.get(KeyFactory.stringToKey(orderKeyStr));
+    return order.getProperty(propertyName);
+  }
+
+  /** 
+   * Updates the status property of orders with the keyString from orderKeys list to status.
+   * TODO[ak47na]: use transactions to update the state of orders.
+   */
+  public void updateStatusForOrders(List<String> orderKeys, String status) throws EntityNotFoundException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java]: No tests?

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 15]
```diff
@@ -152,16 +155,40 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
...
+    for (Entity order : results) {
+      resultsKeyStrings.add(KeyFactory.keyToString(order.getKey()));
+    }
+    return resultsKeyStrings;
+  }
+
+  /** 
+   * Returns the property propertyName of order with keyString representation orderKeyStr.
+   */
+  public Object getProperty(String orderKeyStr, String propertyName) throws EntityNotFoundException {
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java]: No tests?

File: src/test/java/com/google/sps/data/OrderHandlerTest.java [hunk 16]
```diff
@@ -137,19 +138,38 @@ public void testTwoSuccesfulOrders() throws InterruptedException, ApiException,
...
+    }
+    return resultsKeyStrings;
+  }
+
+  public void testGetAvailableOrders() throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
+    inintializeBooksDatastore();
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    orderHandler.makeOrders("user1", new Point(3, 3), Arrays.asList("book1", "book2", "book3"));
+    orderHandler.makeOrders("user1", new Point(3, 3), Arrays.asList("book1", "book3"));
+    List<String> expected = getEntitiesKeyString(ds.prepare(new Query("Order").
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/OrderHandlerTest.java]: You should generally not rely on `makeOrders` to test `getAvailableOrders`. You should also not populate the `expected` variable with code which is exactly how `getAvailableOrders` is implemented.

Let's assume that makeOrders has a bug which makes it so that it doesn't actually write anything to datastore. This test would pass regardless of the implementation since `expected` will be empty. And yes you can argue that `makeOrders` itself is properly tested (or maybe it's also tested via getAvailableOrders, which masks the same problem).

A simpler way to write this test would be to directly write something to the datastore and then verify that get can correctly retrieve it:

first_order = datastore.put(foo);
second_order = datastore.put(bar);
Assert.assertEquals({first_order.getKey(), second_order.getKey()}, orderHandler.getAvailableOrders(...);

File: src/main/java/com/google/sps/data/Journey.java (hunk 8 above)
Comment (alexcepoi) [reply src/main/java/com/google/sps/data/Journey.java]: It makes sense now. I don't know why I was confused. I think I was looking for something which explains why config = 0 should not be updated. e.g.
Start with config = 1, since config=0 means no points should be visited, so nothing to do.

File: src/main/java/com/google/sps/data/Journey.java [hunk 17]
```diff
@@ -56,12 +59,235 @@ public Pair(int first, int second) {
...
+   *
+   * The method will call initializeBestTimeMatrix() to set the initial states for bestTime.
+   * for each configuration config from 1 to (1<<(numberOfWaypoints - 1)) - 1, AND 
+   *   for each waypointIndex i from 0 to numberOfWaypoints - 1 that is not set to 1 in config
+   *   AND there is no index k set to 1 in config such that waypoint k must be visited after i
+   *     newConfig = config + (1<<i), i.e. newConfig has bits in config plus the i-th bit set to 1
+   *     update bestTime[newConfig][i] with each matrix element bestTime[config][j] for which
+   *     j is set to 1 in config, i.e. consider each index j set to 1 in config as the last visited
+   *     waypoint before i and update bestTime[newConfig][i].first with bestTime[config][j].first +
+   *     the time to get from waypoint j to waypoint i.
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java:191]: This might be just me, but phrase completely went over my head. I think it's just saying the same thing as the expressions below (which were much easier for me to digest for some reason) + the fact that it respects point restrictions?

File: src/main/java/com/google/sps/data/Journey.java [hunk 18]
```diff
@@ -56,12 +59,235 @@ public Pair(int first, int second) {
...
+    return bestTime;
+  }
+
+  /**
+   * Computes and returns the bestTime matrix such that bestTime[config][latsWaypoint].first is the
+   * min time needed to visit all points with bits set to 1 in config such that the path ends in 
+   * lastWaypoint and bestTime[config][latsWaypoint].second represents the point that was visited 
+   * right before lastWaypoint. 
+   *
+   * The method will call initializeBestTimeMatrix() to set the initial states for bestTime.
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/Journey.java]: The fact that this method will call another is an implementation detail. You shouldn't document things like this in the contract of the method (and the user doesn't really care that this calls a private method anyway).

What you should document is behavior, which you also do (e.g. This will return a matrix where bla bla) and any caveats (e.g. it will reutrn (MAX_VALUE, -1) for row 0).

What happens if there is no route which respects the provided restrictions, what will each element be set to? Also (MAX_VALUE, -1)?

File: src/test/java/com/google/sps/data/JourneyTest.java [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+/**
+ * Tests the behaviour of a Journey object.
+ * TODO[ak47na]: Implement brute-force algorithm that iterates over all permutations of waypoints
+ * and finds the optimal one.
+ */
+@RunWith(JUnit4.class)
+public class JourneyTest {
+  private ManhattanDistancePathFinder pathFinder = new ManhattanDistancePathFinder();
+  @Test
+  public void testOptimalTimeJourney1() throws ApiException, BadRequestException, DataNotFoundException, IOException, InterruptedException {
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/JourneyTest.java]: Try to work out in the test name how does testOptimalTimeJourney1 differ from testOptimalTimeJourney1. In general, each test should be testing a single specific behavior, and you should feel somewhat confident that the code works correctly if it produces the right result.

File: src/test/java/com/google/sps/data/DeliverySystemTest.java [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+  private void createOrder(CourierStop library, CourierStop recipient, List<String> bookIds, String userId, boolean addOrderKey) {
+    String orderKey = orderHandler.addOrderToDatastore((LibraryPoint)library.getPoint(), bookIds, userId, recipient.getPoint());
+    if (addOrderKey) {
+      library.addOrderKey(orderKey);
+      recipient.addOrderKey(orderKey);
+    }
+  }
+
+  @Test
+  public void testTakeAllOrders() throws ApiException, BadRequestException, DataNotFoundException, EntityNotFoundException, IOException, InterruptedException  {
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/DeliverySystemTest.java]: nit: try to work out the conditions in the test names, to make it clear what the test is testing:
TakeAllOrdersWhereSlotIsLargeEnough
TakeNoOrderIfSlotTooSmall
TakeOnlyOneOrderIfSecondOrderExceedsDeliverySlot

**Review by alexcepoi — APPROVED** [review]

File: src/main/java/com/google/sps/data/CourierStop.java (hunk 12 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/CourierStop.java:41]: No, I removed the if statement.

File: src/test/java/com/google/sps/data/JourneyHandlerTest.java (hunk 13 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/JourneyHandlerTest.java:53]: I removed it.

File: src/test/java/com/google/sps/data/JourneyTest.java (hunk 19 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/JourneyTest.java]: I changed the names to `testOptimalTimeJourneyRestrictionsDoNotAffectOptimalOrder` and `testOptimalTimeJourneyRestrictionsAffectOptimalOrder`

File: src/test/java/com/google/sps/data/DeliverySystemTest.java (hunk 20 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/DeliverySystemTest.java]: Done.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 15 above)
Comment (ak47na) [reply src/main/java/com/google/sps/data/OrderHandler.java]: I added one for `getProperty` and one for `updateStatusForOrders`.

File: src/test/java/com/google/sps/data/OrderHandlerTest.java (hunk 16 above)
Comment (ak47na) [reply src/test/java/com/google/sps/data/OrderHandlerTest.java]: It makes sense. I changed the tests to call `addOrderToDatastore()`(a method that only creates and adds the order) instead.
